<a href="https://colab.research.google.com/github/lorenzo-arcioni/programmazione-python-base/blob/main/Capitolo2_Numeri_e_Stringhe/2_Aritmetica.ipynb" target="_blank"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# ➕ Aritmetica in Python

In questo notebook vediamo come Python gestisce le **operazioni matematiche** di base e i comportamenti dei numeri nei calcoli.

Python supporta i classici operatori aritmetici usati in algebra:


| Operatore | Descrizione          | Esempio     | Risultato     |
|-----------|----------------------|-------------|----------------|
| `+`       | Addizione            | `5 + 3`     | `8`            |
| `-`       | Sottrazione          | `10 - 4`    | `6`            |
| `*`       | Moltiplicazione      | `2 * 7`     | `14`           |
| `/`       | Divisione (float)    | `8 / 3`     | `2.666...`     |
| `//`      | Divisione intera     | `8 // 3`    | `2`            |
| `%`       | Modulo (resto)       | `8 % 3`     | `2`            |
| `**`      | Potenza              | `2 ** 3`    | `8`            |


🧠 Python tratta i numeri come **oggetti** e le operazioni matematiche sono in realtà **metodi** definiti sui tipi numerici (`int`, `float`, `complex`, ecc.).

Per effettuare un'operazione in Python, come ad esempio $5 + 3$, ci basta eseguire semplicemente:

In [4]:
5 + 3

8

**Nota Bene**: Nei Jupyter Notebook, l'ultima istruzione di una cella viene stampata automaticamente anche senza `print()`. Questo negli script Python normali non succede, infatti per poter stampare il risultato di un'operazione è necessario utilizzare la funzione `print()`. 

In [5]:
print(5 + 3)

8


Possiamo assegnare il risultato di un'operazione ad una variabile, come ad esempio `somma = 5 + 3`:

In [6]:
somma = 5 + 3
print("Somma:", somma)

Somma: 8


### 🔢 Precedenza degli operatori

Come in matematica, alcuni operatori hanno precedenza su altri:

1. `**`
2. `*`, `/`, `//`, `%`
3. `+`, `-`

Puoi usare le **parentesi `()`** per forzare un ordine specifico.

In [7]:
print(2 + 3 * 4)       # 14 → moltiplicazione prima
print((2 + 3) * 4)     # 20 → parentesi prima
print(3 * 5 % 2)       # 1  → moltiplicazione prima
print(3 * (5 % 2))     # 3  → parentesi prima

14
20
1
3


## 🔢 Comportamento dei Tipi Numerici

Python gestisce automaticamente il tipo di numero in base al contesto dell’operazione:

- Se operi con due `int`, ottieni un `int` (salvo divisione `/`)
- Se uno degli operandi è `float`, il risultato sarà `float`
- Le divisioni con `/` **restituiscono sempre un `float`**, anche se il risultato è intero
- La divisione intera `//` **tronca** il risultato verso il basso

In [8]:
# Somma intera
print(4 - 3)

# Moltiplicazione intero-float
print(2 * 3.0)

# Divisione
print(6 / 3) # float

# Divisione intera
print(7 // 2)

1
6.0
2.0
3


## ⚠️ Casi Particolari: Divisione intera `//` e Modulo `%`

Python applica regole precise e coerenti anche con:
- **numeri negativi**
- **valori con la virgola (float)**

### 🔄 Divisione Intera con Float (`//`)

Anche se `//` restituisce un risultato **intero**, il **tipo del risultato** dipende dagli operandi:
- Se entrambi sono `int` → ritorna `int`
- Se almeno uno è `float` → ritorna `float` (senza parte decimale)

In [9]:
print(7 // 2)     # 3 (int)
print(7.0 // 2)   # 3.0 (float)
print(7 // 2.0)   # 3.0

3
3.0
3.0


### 📉 Divisione Intera con Numeri Negativi

Python **tronca verso il basso** (verso l’inf), non verso lo zero:

In [10]:
print(7 // 3)     # 2
print(-7 // 3)    # -3 (non -2)

2
-3


> Questo comportamento è diverso da alcuni altri linguaggi come C o Java, che troncano verso zero.

### 🧮 Modulo con Numeri Negativi

Il risultato del modulo `%` ha sempre il **segno del divisore**, non del dividendo:

In [11]:
print(7 % 3)      # 1
print(-7 % 3)     # 2
print(7 % -3)     # -2
print(-7 % -3)    # -1

1
2
-2
-1


#### Perché il modulo si calcola così?

In Python (e in molti altri linguaggi), il modulo è definito in modo che valga la seguente relazione:

$$
a \% b = a - b \cdot \left\lfloor \frac{a}{b} \right\rfloor
$$

Questa formula garantisce due cose fondamentali:

1. **Consistenza con la divisione intera**:  
   Si mantiene la relazione:
   $$
   a = b \cdot (a // b) + (a \% b)
   $$
   cioè il valore originale `a` è sempre ricostruibile.

2. **Segno del risultato coerente con il divisore `b`**:  
   Il risultato del modulo ha lo stesso segno di `b`, il che è utile in molti contesti (come nel calcolo di indici ciclici).

---

#### 🔍 Dimostrazione: il modulo ha lo stesso segno del divisore `b`

Partiamo dalla definizione del modulo in Python:

$$
a \% b = a - b \cdot \left\lfloor \frac{a}{b} \right\rfloor
$$

Ora consideriamo il segno del risultato. Chiamiamo:

- $q = \left\lfloor \frac{a}{b} \right\rfloor$ (quoziente intero)
- $r = a - b \cdot q$ (resto, cioè il modulo)

Osserviamo che:

$$
r = a \% b = a - bq
$$

Poiché $q = \left\lfloor \frac{a}{b} \right\rfloor$, significa che:

$$
q \leq \frac{a}{b} < q + 1
$$

Moltiplichiamo tutti i membri per $b$ (attenzione al segno di $b$):

- Se $b > 0$:

$$
bq \leq a < b(q + 1)
$$

Sottraendo $bq$ da tutti i membri:

$$
0 \leq a - bq < b \Rightarrow 0 \leq r < b
$$

Quindi:

- Il **modulo `r` è compreso tra 0 (incluso) e `b` (escluso)**.
- E dunque **ha lo stesso segno di `b`**.

- Se $b < 0$:

Stesso procedimento:

$$
bq \geq a > b(q + 1)
$$

Sottraendo $bq$:

$$
0 \geq a - bq > b \Rightarrow b < r \leq 0
$$

Quindi:

- Il **modulo `r` è compreso tra `b` (escluso) e 0 (incluso)**.
- Anche qui, il **segno di `r` è uguale a quello di `b`**.

### ✅ Conclusione

In tutti i casi, il valore di:

$$
a \% b = a - b \cdot \left\lfloor \frac{a}{b} \right\rfloor
$$

produce un **resto `r` con lo stesso segno di `b`**, e un valore **compreso tra `0` e `|b|`**, oppure tra `-|b|` e `0`, a seconda del segno del divisore.

---

#### Esempio: floor della divisione intera

Consideriamo:

$$
\frac{-7}{3} = -2.\overline{3}
$$

Il **floor** (massimo intero minore o uguale) di questo valore è:

$$
\left\lfloor \frac{-7}{3} \right\rfloor = -3
$$

Quindi, nella divisione intera:

$$
-7 \div 3 = -3 \quad \text{(divisione intera con floor)}
$$

E il modulo corrispondente è:

$$
-7 \bmod 3 = -7 - 3 \cdot (-3) = -7 + 9 = 2
$$

✔️ Il risultato ha **lo stesso segno del divisore**, come previsto in Python.

Quindi, in Python:

```python
a % b == a - b * (a // b)
```

### 🔬 Modulo e Divisione Intera con Float

Entrambe le operazioni sono **consentite con numeri decimali**, ma possono generare risultati inaspettati se non si è consapevoli del funzionamento.

#### Divisione intera con float

Quando si esegue la divisione intera `a // b` con numeri decimali (float), Python calcola:

$$
a // b = \left\lfloor \frac{a}{b} \right\rfloor
$$

ma il risultato è comunque un numero in **virgola mobile**.

**Esempio:**

$$
7.5 // 2 = \left\lfloor 3.75 \right\rfloor = 3.0
$$

$$
-7.5 // 2 = \left\lfloor -3.75 \right\rfloor = -4.0
$$

#### Modulo con float

Il modulo `a % b` con float è calcolato usando la stessa formula generale:

$$
a \% b = a - b \cdot \left\lfloor \frac{a}{b} \right\rfloor
$$

**Esempi:**

- Positivo:

$$
7.5 \% 2 = 7.5 - 2 \cdot \left\lfloor 3.75 \right\rfloor = 7.5 - 2 \cdot 3 = 1.5
$$

- Negativo:

$$
-7.5 \% 2 = -7.5 - 2 \cdot \left\lfloor -3.75 \right\rfloor = -7.5 - 2 \cdot (-4) = -7.5 + 8 = 0.5
$$

✔️ Anche con i float, il risultato del modulo ha **lo stesso segno del divisore `b`**.

In [12]:
print(7.5 // 2.0)   # 3.0
print(7.5 % 2.0)    # 1.5

print(-7.5 // 2.0)  # -4.0
print(-7.5 % 2.0)   # 0.5

3.0
1.5
-4.0
0.5


📌 **Regola generale**:
> Per ogni coppia `a`, `b` vale sempre la relazione:  
>  `a == (a // b) * b + (a % b)`

In [13]:
a = -7.5
b = 2.0
print(a == (a // b) * b + (a % b))  # True

True


📎 Questo comportamento è utile in algoritmi dove serve controllare **ripetizioni cicliche**, **allineamenti**, **calendari**, ecc.

## 🔢 Arrotondamenti e Conversione a intero

Python offre diversi strumenti per **gestire i numeri decimali**, a seconda di cosa vuoi ottenere:

| Funzione        | Descrizione                                 | Comportamento                       |
|----------------|----------------------------------------------|-------------------------------------|
| `round(x)`      | Arrotonda al numero intero più vicino        | Valore tondo più vicino (con .5 arrotonato verso il pari) |
| `round(x, n)`   | Arrotonda a `n` cifre decimali               | Utile per controllare la precisione |
| `int(x)`        | Converte a intero **troncando** la parte decimale | Non arrotonda, taglia!             |
| `abs(x)`        | Restituisce il valore assoluto del numero x | Toglie il segno al numero |

⚠️ Attenzione:
- `int(3.9)` → `3` (non 4!)
- `round(2.5)` → `2`, perché segue l'**arrotondamento bancario** (verso il numero pari)

In [14]:
print("round(3.14159):", round(3.14159))             # 3
print("round(3.14159, 2):", round(3.14159, 2))       # 3.14
print("int(3.99):", int(3.99))                       # 3
print("round(2.5):", round(2.5))                     # 2 (verso il pari)
print("round(3.5):", round(3.5))                     # 4
print(abs(3))                                        # 3
print(abs(-3))                                       # 3

round(3.14159): 3
round(3.14159, 2): 3.14
int(3.99): 3
round(2.5): 2
round(3.5): 4
3
3


## ⚖️ Perché `round()` in Python arrotonda al numero pari?

Quando usi `round(x)`, Python applica una regola particolare chiamata:

### 🔹 "Round Half To Even"
Conosciuto anche come **arrotondamento bancario**, è il metodo predefinito usato da `round()` per **ridurre l’errore cumulativo** nei calcoli numerici.

### 🧠 Come funziona?
Se un numero è esattamente a metà tra due interi (es. `2.5`, `3.5`, `4.5`...), Python **arrotonda verso il numero pari più vicino**:

| Numero        | Risultato `round(x)` |
|---------------|----------------------|
| `2.5`         | `2` ← pari            |
| `3.5`         | `4` ← pari            |
| `4.5`         | `4` ← pari            |
| `5.5`         | `6` ← pari            |

📌 Ma se **non è a metà esatta**, il numero viene arrotondato normalmente al più vicino:

| Numero        | Risultato `round(x)` |
|---------------|----------------------|
| `2.4`         | `2`                  |
| `2.6`         | `3`                  |

### ✅ Perché si fa così?
L’arrotondamento al pari è usato per **evitare distorsioni sistematiche** nei dati:
- Se si arrotondasse sempre per eccesso (`.5` → su), si introdurrebbe **un piccolo bias positivo** in grandi somme o medie.
- Arrotondare al pari **bilancia verso l’alto e verso il basso** in modo equo.

Questo metodo è adottato anche in linguaggi come Java, R, C# e nei fogli di calcolo (es. Excel, con `=ROUND()`).

In [15]:
print(round(2.5))  # 2
print(round(3.5))  # 4
print(round(4.5))  # 4
print(round(5.5))  # 6

2
4
4
6


## 🧮 Funzioni matematiche utili in Python

Python fornisce alcune **funzioni matematiche di base** integrate (senza moduli esterni):

- `abs(x)` → valore assoluto
- `pow(x, y)` → elevamento a potenza (`x ** y`)
- `divmod(x, y)` → restituisce una **tupla** `(x // y, x % y)` (divisione intera + resto)

In [16]:
print("abs(-10):", abs(-10))           # 10
print("pow(2, 3):", pow(2, 3))         # 8
print("2 ** 3:", 2 ** 3)               # 8 (equivalente)
print("divmod(17, 5):", divmod(17, 5)) # (3, 2)

abs(-10): 10
pow(2, 3): 8
2 ** 3: 8
divmod(17, 5): (3, 2)


`divmod()` è molto utile in operazioni dove servono **entrambi** i risultati della divisione: quoziente e resto.

## 🧠 Numeri Complessi

Python supporta nativamente i **numeri complessi**, usati in matematica, ingegneria, fisica, ecc.

- Un numero complesso è del tipo `a + bj`, dove:
  - `a` è la parte reale
  - `b` è la parte immaginaria
- Si crea con la lettera `j` (non `i`!)

In [17]:
c1 = 3 + 4j
c2 = 1 - 2j

print("Somma:", c1 + c2)
print("Prodotto:", c1 * c2)
print("Parte reale di c1:", c1.real)
print("Parte immaginaria di c1:", c1.imag)
print("Modulo di c1:", abs(c1))  # √(a² + b²)

Somma: (4+2j)
Prodotto: (11-2j)
Parte reale di c1: 3.0
Parte immaginaria di c1: 4.0
Modulo di c1: 5.0


## ⚠️ Limiti della rappresentazione dei numeri in virgola mobile (`float`)

In Python (come in quasi tutti i linguaggi), i numeri decimali (`float`) sono rappresentati **in virgola mobile** secondo lo standard **IEEE 754 a 64 bit**.

### 🧠 Cosa significa?
- Solo una quantità **finita** di numeri può essere rappresentata in modo **esatto**
- Alcuni numeri, come `0.1`, **non si possono rappresentare esattamente**
- Si introducono **approssimazioni**, che possono causare **piccoli errori nei calcoli**


### 🔎 Esempi pratici:
#### 1. **Errore di rappresentazione**

In [18]:
# Errore di rappresentazione
print("0.1 + 0.2 =", 0.1 + 0.2)  # Non sarà 0.3 esatto

0.1 + 0.2 = 0.30000000000000004


📉 **Risultato**: `0.30000000000000004`

Anche se **matematicamente** `0.1 + 0.2 = 0.3`, in binario questi numeri hanno una rappresentazione **infinita**, che viene **troncata**. L’errore è minimo, ma c’è!

#### 2. **Overflow** (numero troppo grande)

In [19]:
# Precisione persa con numeri molto grandi
print("1e308 * 10 =", 1e308 * 10)  # Inf (overflow)

1e308 * 10 = inf


📉 **Risultato**: `inf` (infinito)

Il valore `1e308` è già vicino al massimo rappresentabile (~`1.8e308`). Moltiplicandolo per 10, si supera il limite → Python restituisce `inf`.

#### 3. **Underflow** (numero troppo piccolo)

In [20]:
# Sotto il minimo rappresentabile
print("1e-324 / 10 =", 1e-324 / 10)  # 0.0 (underflow)

1e-324 / 10 = 0.0


📉 **Risultato**: `0.0`

Il numero `1e-324` è **il più piccolo numero positivo** rappresentabile in `float`. Dividendo ancora, il risultato diventa **così vicino a zero** che non può più essere distinto da zero stesso → Python restituisce `0.0`.

## 🔁 Assegnazioni combinate

In Python puoi usare **operatori abbreviati** per modificare il valore di una variabile direttamente, invece di riscrivere l'intera espressione.

### 📌 Esempi comuni:

| Operatore | Equivalente     | Effetto                       |
|-----------|------------------|-------------------------------|
| `+=`      | `x = x + y`      | somma e assegna               |
| `-=`      | `x = x - y`      | sottrai e assegna             |
| `*=`      | `x = x * y`      | moltiplica e assegna          |
| `/=`      | `x = x / y`      | dividi (float) e assegna      |
| `//=`     | `x = x // y`     | dividi intera e assegna       |
| `%=`      | `x = x % y`      | assegna il resto della divisione |
| `**=`     | `x = x ** y`     | eleva a potenza e assegna     |

Questi operatori rendono il codice più **conciso e leggibile**, specialmente nei cicli o nei contatori.

In [21]:
x = 10
x += 5     # x = x + 5 → 15
print(x)

x *= 2     # x = x * 2 → 30
print(x)

x %= 7     # x = x % 7 → 2
print(x)

15
30
2


## ➕➖ Operatori unari in Python

Python supporta anche gli **operatori unari**:

- `+x`: restituisce il valore **positivo** di `x` (in pratica, non cambia nulla)
- `-x`: restituisce l'**opposto** di `x` (lo cambia di segno)

> ❗ Non esistono operatori `++` o `--` come in C/C++ o Java. Scrivere `x++` dà errore.

### 🧠 Doppi operatori (`--x`, `+-x`, ecc.)
Python valuta da **sinistra a destra**:
- `--x` significa `-(-x)` → diventa **positivo**
- `+-x` significa `+(-x)` → resta **negativo**

Sono legittimi e possono essere usati, ma raramente sono necessari.

In [22]:
x = 5
print(+x)     # 5
print(-x)     # -5

print(--x)    # 5 → -(-5)
print(+-x)    # -5 → +(-5)

# Errore: Python non ha ++
# x++  ❌ SyntaxError

5
-5
5
-5


## 📚 Rappresentazione di Numeri in Diverse Basi

Python permette di rappresentare numeri interi in diverse **basi numeriche** grazie a dei **prefissi**:

- `0b` → **binario** (base 2), ad esempio `0b1010`
- `0o` → **ottale** (base 8), ad esempio `0o17`
- `0x` → **esadecimale** (base 16), ad esempio `0x1A`

Questi valori vengono automaticamente convertiti in **decimale** quando usati nel codice. È utile per lavorare con bit, permessi Unix, colori HTML e altro ancora.

Esempio:
- `0b1010` → 10
- `0o17` → 15
- `0x1A` → 26

Puoi usare questi numeri come normali interi in qualsiasi operazione aritmetica.

In [23]:
# Esempi di numeri in diverse basi
binario = 0b10_10       # 10 in base 10
ottale = 0o17           # 15 in base 10
esadecimale = 0x1A      # 26 in base 10
decimale = 100          # già in base 10

print("Binario (0b1010)      →", binario)
print("Ottale (0o17)         →", ottale)
print("Esadecimale (0x1A)    →", esadecimale)
print("Decimale (100)        →", decimale)

Binario (0b1010)      → 10
Ottale (0o17)         → 15
Esadecimale (0x1A)    → 26
Decimale (100)        → 100


## 🔢 Conversioni tra basi numeriche

In Python puoi convertire numeri interi nelle **diverse basi**: binaria, ottale, esadecimale.

### 🔄 Funzioni utili:
- `bin(x)` → restituisce la **rappresentazione binaria** (`0b...`)
- `oct(x)` → restituisce l'**ottale** (`0o...`)
- `hex(x)` → restituisce l'**esadecimale** (`0x...`)
- `int(stringa, base)` → converte da **qualsiasi base a decimale**

### 🧠 Le stringhe di ritorno includono il prefisso (`0b`, `0o`, `0x`).

In [24]:
# Da decimale ad altre basi
x = 42
print(bin(x))   # '0b101010'
print(oct(x))   # '0o52'
print(hex(x))   # '0x2a'

# Da stringa di base N a intero decimale
print(int("101010", 2))   # 42 (binario → decimale)
print(int("2a", 16))      # 42 (esadecimale → decimale)
print(int("52", 8))       # 42 (ottale → decimale)

0b101010
0o52
0x2a
42
42
42


In [2]:
import nbformat

with open("2_Aritmetica.ipynb", "r", encoding="utf-8") as f:
    notebook = nbformat.read(f, as_version=4)

with open("solo_markdown.md", "w", encoding="utf-8") as out:
    for cell in notebook.cells:
        if cell.cell_type == "markdown":
            out.write(cell.source + "\n\n")

## 👀 Operatori di confronto e tipo booleano

Python fornisce diversi **operatori logici e di confronto** per confrontare valori e costruire condizioni. Questi operatori restituiscono sempre un **valore booleano**, cioè:

- `True` (vero)
- `False` (falso)

### 🧪 Tipi di confronto

| Operatore | Descrizione                | Esempio        | Risultato  |
|-----------|----------------------------|----------------|------------|
| `==`      | Uguaglianza                | `5 == 5`     | `True`     |
| `!=`      | Diverso da                 | `5 != 3`     | `True`     |
| `>`       | Maggiore                   | `7 > 4`      | `True`     |
| `<`       | Minore                     | `3 < 2`      | `False`    |
| `>=`      | Maggiore o uguale          | `5 >= 5`     | `True`     |
| `<=`      | Minore o uguale            | `2 <= 3`     | `True`     |

⚠️ **Attenzione**:
- L'operatore di **assegnazione** è `=` (non confonderlo con `==`).
- L'operatore `<>` non è più valido in Python 3 (era un'alternativa a `!=` in Python 2).

In [3]:
# Esempi di operatori di confronto
print("5 == 5:", 5 == 5)
print("5 != 3:", 5 != 3)
print("7 > 4:", 7 > 4)
print("3 < 2:", 3 < 2)
print("5 >= 5:", 5 >= 5)
print("2 <= 3:", 2 <= 3)

5 == 5: True
5 != 3: True
7 > 4: True
3 < 2: False
5 >= 5: True
2 <= 3: True


### 🧠 Il tipo `bool`

Il tipo `bool` rappresenta i **valori logici**. È una sottoclasse del tipo `int`:
- `True` è rappresentato come `1`
- `False` come `0`

Puoi usare la funzione `bool()` per convertire un valore in booleano:

- `bool(0)` → `False`  
- `bool(1)` → `True`  
- `bool("")` → `False` (stringa vuota)  
- `bool("ciao")` → `True` (stringa non vuota)  
- `bool([])` → `False` (lista vuota)  
- `bool([1, 2])` → `True` (lista non vuota)

📌 In generale, **sono considerati falsi**:
- `False`
- `None`
- `0, 0.0`
- sequenze vuote: `'', [], {}, set()`

Tutti gli altri valori sono considerati `True`.

In [4]:
# Conversioni a booleano
print("bool(0):", bool(0))
print("bool(1):", bool(1))
print("bool(\"\"):", bool(""))
print("bool(\"ciao\"):", bool("ciao"))
print("bool([]):", bool([]))
print("bool([1, 2]):", bool([1, 2]))

# Tipi considerati "falsi"
falsi = [False, None, 0, 0.0, '', [], {}, set()]
print("Tutti i seguenti sono valutati come False:")
for val in falsi:
    print(f"{repr(val)} -> {bool(val)}")

bool(0): False
bool(1): True
bool(""): False
bool("ciao"): True
bool([]): False
bool([1, 2]): True
Tutti i seguenti sono valutati come False:
False -> False
None -> False
0 -> False
0.0 -> False
'' -> False
[] -> False
{} -> False
set() -> False


### 🔗 Confronti multipli concatenati

Python permette di concatenare più confronti in modo leggibile:

- `3 < 5 < 10` → `True` (equivale a `3 < 5 and 5 < 10`)  
- `x == y == z` → `True` solo se tutti e tre i valori sono uguali

In [5]:
# Confronti concatenati
print("3 < 5 < 10:", 3 < 5 < 10)
print("3 < 5 and 5 < 10:", 3 < 5 and 5 < 10)

x, y, z = 5, 5, 5
print("x == y == z:", x == y == z)

a, b, c = 1, 2, 3
print("a < b < c:", a < b < c)

3 < 5 < 10: True
3 < 5 and 5 < 10: True
x == y == z: True
a < b < c: True


### ⚖️ Confronti tra tipi diversi

Python può confrontare tipi **compatibili** (es. `int` e `float`), ma restituisce `False` se i valori sono diversi:

- `5 == 5.0` → `True` (int e float compatibili)  
- `5 == '5'` → `False` (int e stringa non confrontabili)

In [6]:
# Tipi diversi
print("5 == 5.0:", 5 == 5.0)    # True, int e float compatibili
print("5 == '5':", 5 == '5')    # False, int e stringa non compatibili
print("5 != '5':", 5 != '5')    # True

5 == 5.0: True
5 == '5': False
5 != '5': True


### 🧠 Uguaglianza vs Identità

- `==` → confronta **i valori**
- `is` → confronta **l'identità degli oggetti** (cioè se puntano alla stessa area di memoria)

Esempio:

- `a = [1, 2, 3]`  
- `b = [1, 2, 3]`  
- `a == b` → `True` (stessi valori)  
- `a is b` → `False` (oggetti distinti)

In [7]:
a = [1, 2, 3]
b = [1, 2, 3]
c = a

print("a == b:", a == b)  # True, stessi valori
print("a is b:", a is b)  # False, oggetti diversi
print("a is c:", a is c)  # True, stessa identità

a == b: True
a is b: False
a is c: True


### ℹ️ Conversione implicita `bool` → `int`

Poiché `bool` deriva da `int`, puoi usarli nei calcoli:

- `True + True` → `2`  
- `False * 5` → `0`  
- `int(True)` → `1`  
- `int(False)` → `0`

📌 Anche le espressioni booleane possono essere usate in operazioni matematiche (ma con cautela!).

In [8]:
print("True + True:", True + True)
print("False * 5:", False * 5)
print("int(True):", int(True))
print("int(False):", int(False))

# Uso in espressioni
a = True
b = False
print("a + 3:", a + 3)   # 1 + 3 = 4
print("b + 3:", b + 3)   # 0 + 3 = 3

True + True: 2
False * 5: 0
int(True): 1
int(False): 0
a + 3: 4
b + 3: 3


## ✅ Conclusioni

In questo notebook abbiamo esplorato in modo approfondito come Python gestisce le **operazioni aritmetiche** e il comportamento dei numeri.

Abbiamo visto:

- Come usare gli **operatori matematici** (`+`, `-`, `*`, `/`, `//`, `%`, `**`)
- Le **assegnazioni combinate** (`+=`, `*=` ecc.)
- Gli **operatori unari** (`+x`, `-x`, `--x`, ecc.)
- La gestione dei **numeri complessi**
- L’arrotondamento e le funzioni utili come `round()`, `abs()`, `divmod()`, `pow()`
- Le **conversioni tra basi numeriche** (binaria, ottale, esadecimale)
- I limiti della rappresentazione numerica: **overflow, underflow, e precisione**
- Gli operatori booleani e le operazioni booleane.

💡 Python tratta i numeri come **oggetti** e consente di eseguire calcoli in modo molto flessibile, ma è importante conoscere le **caratteristiche tecniche** per evitare errori nei conti o risultati inattesi.

➡️ **Prosegui con il prossimo notebook sulle stringhe** per imparare a manipolare testi e caratteri in Python:
<a href="https://colab.research.google.com/github/lorenzo-arcioni/programmazione-python-base/blob/main/Capitolo2_Numeri_e_Stringhe/3_Stringhe.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>